src/type_object.rs

```diff
@@ -64,21 +66,71 @@ pub unsafe trait PyTypeInfo: Sized + HasPyGilRef {
 
     /// Returns the safe abstraction over the type object.
     #[inline]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyTypeInfo::type_object` will be replaced by `PyTypeInfo::type_object_bound` in a future PyO3 version"
+        )
+    )]
     fn type_object(py: Python<'_>) -> &PyType {
         unsafe { py.from_borrowed_ptr(Self::type_object_raw(py) as _) }
     }
 
+    /// Returns the safe abstraction over the type object.
+    #[inline]
+    fn type_object_bound(py: Python<'_>) -> Bound<'_, PyType> {
```

Comment:
I considered whether to make `type_object_bound`, `is_type_of_bound` and `is_exact_type_of_bound` have defaults implemented in terms of their original methods (or vice versa).

Given that we don't expect users to be implementing `PyTypeInfo`, I decided it was probably simpler to just give them all defaults which match their final expected forms.

Reply:
I agree that the new methods should be implemented in their final form. Whats the reason not reimplementing the old API in terms of the new one?

Reply:
That's a very good point, I will go ahead and do that 👍

Reply:
Ah, there is one reason: `type_object_bound` now returns an owned `Bound` instead of a borrowed gil-ref, which I think is correct for the safety reasoning I express in the note below.

But `is_type_of` and `is_exact_type_of` can be implemented in terms of their new replacements 👍